projects/geometry/101_color_cube_metal/101_color_cube_metal.cpp

```diff
@@ -0,0 +1,265 @@
+        if (pError != nullptr) {                                                     \
+            std::stringstream ss;                                                    \
+            ss << "\n";                                                              \
+            ss << "*** FUNCTION CALL FAILED *** \n";                                 \
+            ss << "FUNCTION: " << #FN << "\n";                                       \
+            ss << "Error: " << pError->localizedDescription()->utf8String() << "\n"; \
+            ss << "\n";                                                              \
+            GREX_LOG_ERROR(ss.str().c_str());                                        \
+            assert(false);                                                           \
+        }                                                                            \
+    }
+
+// =============================================================================
+// Shader code
+// =============================================================================
+const char* gShaders = R"(
+	#include <metal_stdlib>
+	using namespace metal;
+
+	struct Camera {
+		float4x4 MVP;
+	};
+
+	struct VSOutput {
+		float4 PositionCS [[position]];
+		float3 Color;
+	};
+
+	struct VertexData {
+		float3 PositionOS [[attribute(0)]];
+		float3 Color [[attribute(1)]];
+	};
+
+   VSOutput vertex vertexMain(
+		VertexData vertexData [[stage_in]],
+		constant Camera &Cam [[buffer(2)]])
+	{
+		VSOutput output;
+		float3 position = vertexData.PositionOS;
+		output.PositionCS = Cam.MVP * float4(position, 1.0f);
+		output.Color = vertexData.Color;
+		return output;
+	}
+
+	float4 fragment fragmentMain( VSOutput in [[stage_in]] )
+	{
+		return float4(in.Color, 1.0);
+	}
+)";
+
+// =============================================================================
+// Globals
+// =============================================================================
+static uint32_t gWindowWidth  = 1280;
+static uint32_t gWindowHeight = 720;
+static bool     gEnableDebug  = true;
+
+void CreateGeometryBuffers(
+    MetalRenderer* pRenderer,
+    MetalBuffer*   pIndexBuffer,
+    MetalBuffer*   pPositionBuffer,
+    MetalBuffer*   pVertexColorBuffer);
+
+// =============================================================================
+// main()
+// =============================================================================
+int main(int argc, char** argv)
+{
+    std::unique_ptr<MetalRenderer> renderer = std::make_unique<MetalRenderer>();
+
+    if (!InitMetal(renderer.get(), gEnableDebug)) {
+        return EXIT_FAILURE;
+    }
+
+    // *************************************************************************
+    // Compile shaders
+    // *************************************************************************
+    MetalShader vsShader;
+    MetalShader fsShader;
+    {
+        NS::Error*    pError   = nullptr;
+        MTL::Library* pLibrary = renderer->Device->newLibrary(
```

Comment:
Seems like a possible case for NS::SharedPtr?

Reply:
I don't think we need it? I believe the AutoreleasePool will guarantee that nothing gets deleted within it, and we don't need the pLibrary pointer after the shaders are compiled. At least, I don't think we use it later (that would be a bug and require an NS::SharedPtr for sure).